### DISCUSSION TITLE
Rewrite Kapitan as Ansible roles and/or plugins?

### DISCUSSION THREAD

**Discussion by a-prokopyev-resume** [deepmind/kapitan#1193] [General]
Hello,

I have recently discovered your `Kapitan` project for myself and I think the ideas and whole conception of the project are brilliant and relatively unique taking into account all its features.

But current realization of `Kapitan` IMHO is too buggy unfortunately.

Why would not implement the same idea as modules of some other big established `IaaC` tool like `Ansible`, `Salt`, etc. ?

The install from your latest docker image even cannot execute `kapitan inventory`, it just crashes with an error even on an almost empty project created by the same `kapitan init`.

`kapitan compile` generates expected output files, but even for a such small project it displays some nasty warning most likely due to some bugs in the code base of the `Kapitan`.

**Comment by ademariag** [comment]
Hello @a-prokopyev-resume,

thank you for taking some time to write this conversation, I appreciate it.

I will address each statement individually:

> I have recently discovered your Kapitan project for myself and I think the ideas and whole conception of the project are brilliant and relatively unique taking into account all its features. 

Thank you, it is truly unique, and glad you are able to see it in spite of your relative non pleasant initial experience. Please make sure you join us on the Kubernetes slack we have #kapitan (see https://kapitan.dev/support/)


> But current realization of Kapitan IMHO is too buggy unfortunately.
> The install from your latest docker image even cannot execute kapitan inventory, it just crashes with an error even on an almost empty project created by the same kapitan init.

I can reassure you that Kapitan is working like clockwork for many organizations, and I hardly see "bugs" in my interpretation of "unintended behaviour"

What happens is that, by (ancient) choice, we handle errors by printing the python stack trace, which is somehow not as pleasant, but help debugging more complex scenarios. 

The reason behind it is that usually most bugs will be in the "template" or "generators/inputs" code, rather than the `kapitan` code itself, and so sending the stack trace makes it easier for people to debug it.

However it is especially unpleasant for first time users, and I will take your comment into account and see if I can make it a little more pleasant.

**Reply by a-prokopyev-resume** [reply]
> > But current realization of Kapitan IMHO is too buggy unfortunately.
> > The install from your latest docker image even cannot execute kapitan inventory, it just crashes with an error even on an almost empty project created by the same kapitan init.
> 
> I can reassure you that Kapitan is working like clockwork for many organizations, and I hardly see "bugs" in my interpretation of "unintended behaviour"

Hello,

I used following `bash` aliases:
```
alias kapitan='docker exec -ti -e GIT_PYTHON_REFRESH=0 --workdir /mnt/ kapitan kapitan'
alias kapitan-rm='docker stop kapitan; docker rm kapitan; docker ps -a'
alias kapitan-start='docker run -d -v $(pwd):/mnt:delegated --name kapitan --entrypoint=/bin/ash projectsyn/kapitan:v0.29.5 -c "sleep 10h"; docker ps -a'
alias kapitan2='docker run -t --rm --user root -v $(pwd):/src:delegated kapicorp/kapitan'
```
Command line:    
` set -x; rm -Rf /utils/kapitan/test/*; ls; kapitan2 init; ls; kapitan2 inventory; kapitan inventory | head -n 3`

Please let me know, how can you explain a crash in the following very simple test of the latest docker container with `kapitan` inside it?

![kapitan_inventory_crash](https://github.com/user-attachments/assets/42e6bff7-8a3f-4d74-ba8b-88e3cba3cde4)

And as you can see at the bottom of the screenshot  old version by `Project Syn` works just fine in the same simple test of trying to run `kapitan inventory`.

**Reply by ademariag** [reply]
Hello @a-prokopyev-resume, 

I have been able to replicate the bug, and I will look into it, but most likely it's due to the fact that `kapitan init` is a deprecated way to initialize the repository.

please see https://kapitan.dev/getting_started/

Nevertheless I will be looking at why this results in an error, but most likely it's due to the fact that the files created are not compatible anymore.

**Reply by ademariag** [reply]
Please note that the "cruft" mode is currently affected by https://github.com/orgs/kapicorp/discussions/1193#discussioncomment-10049592

**Comment by ademariag** [comment]
> Why would not implement the same idea as modules of some other big established IaaC tool like Ansible, Salt, etc. ?

I think this suggestion is rooted on a possible misinterpretation of the full potential of `kapitan` which is usually used to generate configuration to support "big established IaaC tool like Ansible, Salt"

I am happy to spend some time to help you get started, however Kapitan is usually a worth experiment if you are handling complex infrastructures or configuration, while if you want to use it for a pet personal project then indeed it might not be the right tool for you.

**Reply by a-prokopyev-resume** [reply]
> I think this suggestion is rooted on a possible misinterpretation of the full potential of `kapitan` which is usually used to generate configuration to support "big established IaaC tool like Ansible, Salt"

What is wrong to automate one popular tool like `Ansible`, `Salt` or `Terraform` from another one like `Ansible` or `Salt`?

My idea was just to offer you to use a more popular, stable and much better tested base/foundation for your software. And then still implement all required features on top of a popular tool like `Ansible` and/or `Salt`.
It would allow you to avoid many bugs already fixed in such foundation and concentrate your work forces just on your very specific feature set.

I guess a universal plugin compatible with  both `Ansible` and `Salt` would be the best.

**Reply by ademariag** [reply]
Thank you for your idea, but this is not an approach we are evaluating at the moment. 

Giving your recent discovery of Kapitan (and the unexpected complications you have experienced) I would recommend you to spend some time with it as it is, before forking it and reimplementing it the way you want.

**Comment by ademariag** [comment]
FYI I am looking at the instructions using the cookiecutter, and I have filed this bug https://github.com/cruft/cruft/issues/305 as I believe this used to work. 

I will then add a test to make sure this doesn't happen anymore
